Improve plugin search panel

takuyakanbr opened the pull request:
**Changes**
- Add a short description for each plugin (when hovering over the plugin name)
- Add tags for each plugin, and use these tags (on top of the name) when searching for plugins
- Add a button to pin/unpin plugins to the top
- Make the search bar stay on top when scrolling through the plugin list
- Add a Back button at the top of the plugin configuration pages
- Allow empty search results

**Screenshots**

Plugin descriptions
![image](https://user-images.githubusercontent.com/875036/41799634-c54af2e8-76a4-11e8-8786-db3cd4030095.png)

Searching for tags
![image](https://user-images.githubusercontent.com/875036/41799640-c70d44a0-76a4-11e8-8df7-31395e79cf6b.png)
![image](https://user-images.githubusercontent.com/875036/41800937-572bb2ba-76aa-11e8-8a62-f25609022c59.png)
![image](https://user-images.githubusercontent.com/875036/41800939-58b678ea-76aa-11e8-858b-1121381a43aa.png)

Pin/unpin plugins + search bar always on top
![image](https://user-images.githubusercontent.com/875036/41799630-c321e4ae-76a4-11e8-868a-223cb15fddc9.png)

Back button
![image](https://user-images.githubusercontent.com/875036/41801093-16c7967a-76ab-11e8-8aee-d4ae0821f140.png)


Empty list if you type nonsense
![image](https://user-images.githubusercontent.com/875036/41799714-10b05f5c-76a5-11e8-947a-f1a0b3ff12ef.png)

Closes #2245
Fixes #1803
Fixes #2234
Fixes #3360
Fixes #3599

jplsek commented:
Awesome work!

Since the summary information of each plugin is now in the PluginDescriptor of the plugin class, and the same information is in the ConfigGroup annotation, would it be possible to just outright remove the ConfigGroup annotation? Thoughts?

deathbeam commented:
Fully agree with @jplsek

takuyakanbr commented:
Added a commit to remove `description` from `ConfigGroup`.

jplsek commented:
The back button at the bottom of the config panel should be removed, since it's now at the top of the config panel.

Also just realized that the tooltip for the title on the individual plugin is gone (I'm guessing from removing the description from the `ConfigGroup` group). I'm not sure how much we care about that. But in general, I want to avoid duplicate names and descriptions, but I don't know how to cleanly solve that issue because of configs that don't have plugins (`RuneLite`, `ChatColor`).
But since `RuneLite`, `ChatColor` are hardcoded anyways, I think `ConfigGroup` should only have the a `keyName` and the plugin name and description should only come from `PluginDescription`, for both the config list, and the individual config. That also allows all the `ConfigGroup`'s to look like `@ConfigGroup("myKeyName")`.

jplsek commented:
I agree with removing the "configuration" from title and adding the toggle button.

I'm not sure about the border because there isn't anywhere else in client that uses borders like that.

Another reason I wanted the back button to be removed is because, based on the discord, people seem to misclick, and click reset lol. But I guess we should get other opinions on this.

Abextm commented:
Rather then adding a border to the top panel you could change it's background color.

takuyakanbr commented:
Added toggle button to config pages and a different background color to the top panel. Also rebased and modified the new plugins accordingly (up to corp beast).

I'm not intending to add any more features to this PR, so feel free to start reviewing.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,248 @@
+	PluginListItem(ConfigPanel configPanel, Plugin plugin, @Nullable Config config,
+		@Nullable ConfigDescriptor configDescriptor)
+	{
+		final PluginDescriptor descriptor = plugin.getClass().getAnnotation(PluginDescriptor.class);
```
This ctor should just call the other one and not duplicate code

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,248 @@
+				configButton.setIcon(CONFIG_ICON);
+			}
+		});
+		configButton.setVisible(true);
```
Add empty line above

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,248 @@
+	private static final ImageIcon OFF_SWITCHER;
+
+	private final ConfigPanel configPanel;
+	private @Getter @Nullable final Plugin plugin;
```
Annotations on new lines always

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -658,4 +467,112 @@ public void focusLost(FocusEvent e)
+	public void onActivate()
+	{
+		super.onActivate();
+		if (searchBar.getParent() != null)
```
Add empty line above

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -658,4 +467,112 @@ public void focusLost(FocusEvent e)
+		}
+	}
+
+	enum DisplayMode
```
Cant this as well be just boolean?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -658,4 +467,112 @@ public void focusLost(FocusEvent e)
+			}
+		}
+
+		if (component instanceof JCheckBox)
```
This can also be just switch. E.g switch component.getClass()

takuyakanbr replied:
I don't think this is possible in Java, unless you mean to use the name of the class, which is possible but probably not a good idea?

deathbeam replied:
Ah right, forgot that Java is bad. Alright :d At least make it chain of if elseif

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -658,4 +467,112 @@ public void focusLost(FocusEvent e)
+
+	private void changeConfiguration(Config config, JComponent component, ConfigDescriptor cd, ConfigItemDescriptor cid)
+	{
+		ConfigItem configItem = cid.getItem();
```
Mark this final

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -162,6 +168,8 @@ public void changedUpdate(DocumentEvent e)
 
 	private void initializePluginList()
 	{
+		List<String> pinnedPlugins = getPinnedPluginNames();
```
Final

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -560,6 +581,27 @@ void stopPlugin(Plugin plugin, PluginListItem listItem)
+
+		if (config == null)
+		{
+			return new ArrayList<>();
```
Collections.emptyList()?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -560,6 +581,27 @@ void stopPlugin(Plugin plugin, PluginListItem listItem)
+			return new ArrayList<>();
+		}
+
+		return Arrays.asList(config.split(","));
```
Use Splitter.on (and make it static final somewhere at the top) and then splitToList(0

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -560,6 +581,27 @@ void stopPlugin(Plugin plugin, PluginListItem listItem)
+
+	void savePinnedPlugins()
+	{
+		String value = pluginList.stream()
```
final

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -560,6 +581,27 @@ void stopPlugin(Plugin plugin, PluginListItem listItem)
+			.filter(PluginListItem::isPinned)
+			.map(PluginListItem::getName)
+			.collect(Collectors.joining(","));
+		configManager.setConfiguration(RUNELITE_GROUP_NAME, PINNED_PLUGINS_CONFIG_KEY, value);
```
empty line above

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -58,20 +58,24 @@
 	private @Nullable final ConfigDescriptor configDescriptor;
 
-	private final String name;
+	private @Getter final String name;
```
Annotations on empty lines always

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -58,20 +58,24 @@
 	private final JLabel toggleButton = new JLabel(OFF_SWITCHER);
 
 	private boolean isPluginEnabled = false;
+	private @Getter boolean isPinned = false;
```
Annotations

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -108,12 +115,19 @@
 	private DisplayMode currentMode = DisplayMode.PLUGIN_LIST;
 	private int scrollBarPosition = 0;
 
+	private JPanel topPanel;
```
Make these 3 final probably

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -262,21 +301,50 @@ private void showMatchingPlugins(Stream<PluginListItem> listItems, String text)
+				topPanelBackButton.setIcon(BACK_ICON);
+			}
+		});
+		topPanelBackButton.setToolTipText("Back");
```
empty line above

takuyakanbr commented:
Made requested changes.

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,300 @@
+		this.name = name;
+		this.description = description;
+		Collections.addAll(keywords, name.toLowerCase().split(" "));
+		Collections.addAll(keywords, tags);
```
`keywords.addAll(tags)`, etc

takuyakanbr replied:
`List#addAll` accepts a `Collection`, not a `String` array. :(

NathenSample replied:
`Arrays.asList()`

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -161,292 +175,179 @@ public void changedUpdate(DocumentEvent e)
-			return;
-		}
+		showMatchingPlugins(pluginList.stream().filter(PluginListItem::isPinned), text);
+		showMatchingPlugins(pluginList.stream().filter(item -> !item.isPinned()), text);
```
Can this be done without passing a Stream around?

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java:
```diff
@@ -161,292 +175,179 @@ public void changedUpdate(DocumentEvent e)
+		topPanelBackButton.addMouseListener(new MouseAdapter()
+		{
+			@Override
+			public void mousePressed(MouseEvent mouseEvent)
```
Prefer ActionListener to MouseAdapter as it works with keyboards. If you must use a MouseAdapter use mouseClicked rather than mousePressed

takuyakanbr replied:
You can't add an `ActionListener` to a `JLabel`. I could go with `mouseClicked`, but the clickables on the other panels (most notably `MaterialTab`s) also uses `mousePressed`, and it will result in an inconsistent experience (one triggers on mouse up and one triggers on mouse down).

It also seems to be harder to trigger a `mouseClicked` on a `JLabel` - if your mouse moves between mouse down and mouse up, it doesn't trigger.

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,300 @@
+			{
+				BufferedImage configIcon = ImageIO.read(ConfigPanel.class.getResourceAsStream("config_edit_icon.png"));
+				CONFIG_ICON = new ImageIcon(configIcon);
+				CONFIG_ICON_HOVER = new ImageIcon(SwingUtil.grayscaleOffset(configIcon, -100));
```
This should be outside of the synchronized block

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,300 @@
+		buttonPanel.add(toggleButton);
+	}
+
+	private void attachPinButtonListener()
```
Do these have to be their own methods?

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,300 @@
+
+	private void attachPinButtonListener()
+	{
+		pinButton.addMouseListener(new MouseAdapter()
```
Prefer ActionListener to MouseAdapter as it works with keyboards. If you must use a MouseAdapter use mouseClicked rather than mousePressed

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,300 @@
+		configButton.addMouseListener(new MouseAdapter()
+		{
+			@Override
+			public void mousePressed(MouseEvent mouseEvent)
```
Prefer ActionListener to MouseAdapter as it works with keyboards. If you must use a MouseAdapter use mouseClicked rather than mousePressed

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java:
```diff
@@ -0,0 +1,300 @@
+		button.addMouseListener(new MouseAdapter()
+		{
+			@Override
+			public void mousePressed(MouseEvent mouseEvent)
```
Prefer ActionListener to MouseAdapter as it works with keyboards. If you must use a MouseAdapter use mouseClicked rather than mousePressed

Abextm commented:
They should be using mouseClicked too.

On Mon, Jun 25, 2018, 06:48 Daniel Teo <notifications@github.com> wrote:

> *@takuyakanbr* commented on this pull request.
> ------------------------------
>
> In
> runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java
> <https://github.com/runelite/runelite/pull/3957#discussion_r197782411>:
>
> >
> -			if (result != JOptionPane.YES_OPTION)
> +		final JLabel topPanelBackButton = new JLabel(BACK_ICON, SwingConstants.LEFT);
> +		topPanelBackButton.setPreferredSize(new Dimension(22, 0));
> +		topPanelBackButton.addMouseListener(new MouseAdapter()
> +		{
> +			@Override
> +			public void mousePressed(MouseEvent mouseEvent)
>
> You can't add an ActionListener to a JLabel. I could go with mouseClicked,
> but the clickables on the other panels (most notably MaterialTabs) also
> uses mousePressed, and it will result in an inconsistent experience (one
> triggers on mouse up and one triggers on mouse down).
>
> It also seems to be harder to trigger a mouseClicked on a JLabel - if
> your mouse moves between mouse down and mouse up, it doesn't trigger.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/runelite/runelite/pull/3957#discussion_r197782411>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALy0PwtF_pdGDtL893yh5Owndjv_3erKks5uANwGgaJpZM4U0e5d>
> .
>

Abextm commented:
Ah yeah, Leave it as is then

On Mon, Jun 25, 2018, 06:49 Daniel Teo <notifications@github.com> wrote:

> *@takuyakanbr* commented on this pull request.
> ------------------------------
>
> In
> runelite-client/src/main/java/net/runelite/client/plugins/config/PluginListItem.java
> <https://github.com/runelite/runelite/pull/3957#discussion_r197783100>:
>
> > +		String name, String description, String... tags)
> +	{
> +		this(configPanel, null, config, configDescriptor, name, description, tags);
> +	}
> +
> +	private PluginListItem(ConfigPanel configPanel, @Nullable Plugin plugin, @Nullable Config config,
> +		@Nullable ConfigDescriptor configDescriptor, String name, String description, String... tags)
> +	{
> +		this.configPanel = configPanel;
> +		this.plugin = plugin;
> +		this.config = config;
> +		this.configDescriptor = configDescriptor;
> +		this.name = name;
> +		this.description = description;
> +		Collections.addAll(keywords, name.toLowerCase().split(" "));
> +		Collections.addAll(keywords, tags);
>
> List#addAll accepts a Collection, not a String array. :(
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/runelite/runelite/pull/3957#discussion_r197783100>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALy0P-VVEanZ9sBBQ3L23AY2Whwz6qIhks5uANx1gaJpZM4U0e5d>
> .
>

takuyakanbr commented:
Made requested changes. For the button listeners, I'm now using `ActionListener`s on a custom `JButton` that draws only an icon.

psikoi commented:
Why was the background color changed? I can't see the before images

takuyakanbr commented:
The background color for the top panel was made slightly lighter to provide visual separation between the top panel and the main panel.

If it's preferred I could undo that, or use a bottom border instead.

With bottom border
![image](https://user-images.githubusercontent.com/875036/42116289-309381da-7c29-11e8-8dfd-c1cc5b848a75.png)
![image3](https://user-images.githubusercontent.com/875036/42116374-9077c6e2-7c29-11e8-8b86-bad755574980.PNG)

Without either

![image2](https://user-images.githubusercontent.com/875036/42116283-2eaa69ba-7c29-11e8-925a-3e8d3453674c.png)
![image4](https://user-images.githubusercontent.com/875036/42116357-7c3fedc6-7c29-11e8-8fde-00bf2636456a.PNG)

psikoi commented:
None of the other scrollable panels have any separation, ge search, farming panels, etc. Was just wondering

Adam- commented:
Whenever I favorite a plugin the scroll jumps weirdly... any idea why?

takuyakanbr commented:
Updated to keep scroll position when pinning/unpinning plugins.
Squashed the new commits into c81cd8c03ca3a0d38c81800564a60afb2e875676.
